core/controllers/ingest.py

```diff
@@ -361,3 +362,30 @@ def remove_unreferenced_organisations():
             db.session.delete(organisation)
 
     db.session.commit()
+
+
+def populate_db_transaction(workbook: dict[str, pd.DataFrame], mappings: tuple[DataMapping]) -> None:
+    """Execute a transaction to populate the database with data from the provided workbook.
+
+    This function wraps the 'populate_db' operation in a database transaction, allowing for
+    retries in case of IntegrityError or InvalidRequestError. If an error occurs during the
+    transaction, the operation is retried up to a specified maximum number of times with a
+    delay of 0.4 seconds between retries.
+
+    :param workbook: A dictionary containing DataFrames representing the workbook data.
+    :param mappings: A tuple of DataMapping instances defining the mapping of workbook data to the database.
+    :return: None
+    """
```

Comment:
We could make this a generic function that accepts a function to call and a set of kwargs to pass to it, as well as parametrising `max_retries` etc. 

Something like:
```
def with_transaction(function: callable, f_kwargs: dict, max_retries: int = 3):
    pass
```
``

Reply:
Good shout on parameterising the function call. We were thinking of setting the max_retries as an env var so that it won't slow down automated unit testing if we decide to increase it by more than three.

Reply:
second this. It would be good to keep context of `populate_db` completely out of the doctring / function scope. So the idea is each function is totally separable in functionality, can be reused for any transaction and tested in isolation from populate_db (although an integration test with both could be useful).

Also, this func catches the specific IntegretyError, would it be appropriate / clear to make the specific exception to be caught also a param?

Reply:
I think I need to investigate more before deciding what to do with the error. There's a possibility another error might arise when queuing up ```populate_db``` like this.